NullPointerException in reproject_and_merge UDF

Tried to write out a parquet file after a raster join, but am getting the following error. I am using version: 0.9.0 RC2.

<details>

---------------------------------------------------------------------------

```

Py4JJavaError                             Traceback (most recent call last)
<ipython-input-16-a57d8cb5c4dd> in <module>
      2     .parquet('/tmp/test.pq',
      3              mode='overwrite',
----> 4              partitionBy='ID'
      5             )

/opt/conda/lib/python3.7/site-packages/pyspark/sql/readwriter.py in parquet(self, path, mode, partitionBy, compression)
    841             self.partitionBy(partitionBy)
    842         self._set_opts(compression=compression)
--> 843         self._jwrite.parquet(path)
    844 
    845     @since(1.6)

/opt/conda/lib/python3.7/site-packages/py4j/java_gateway.py in __call__(self, *args)
   1255         answer = self.gateway_client.send_command(command)
   1256         return_value = get_return_value(
-> 1257             answer, self.gateway_client, self.target_id, self.name)
   1258 
   1259         for temp_arg in temp_args:

/opt/conda/lib/python3.7/site-packages/pyspark/sql/utils.py in deco(*a, **kw)
     61     def deco(*a, **kw):
     62         try:
---> 63             return f(*a, **kw)
     64         except py4j.protocol.Py4JJavaError as e:
     65             s = e.java_exception.toString()

/opt/conda/lib/python3.7/site-packages/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
    326                 raise Py4JJavaError(
    327                     "An error occurred while calling {0}{1}{2}.\n".
--> 328                     format(target_id, ".", name), value)
    329             else:
    330                 raise Py4JError(

Py4JJavaError: An error occurred while calling o621.parquet.
: org.apache.spark.SparkException: Job aborted.
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:198)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand.run(InsertIntoHadoopFsRelationCommand.scala:159)
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:104)
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:102)
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.doExecute(commands.scala:122)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:80)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:80)
	at org.apache.spark.sql.DataFrameWriter$$anonfun$runCommand$1.apply(DataFrameWriter.scala:676)
	at org.apache.spark.sql.DataFrameWriter$$anonfun$runCommand$1.apply(DataFrameWriter.scala:676)
	at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:676)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:285)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:271)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:229)
	at org.apache.spark.sql.DataFrameWriter.parquet(DataFrameWriter.scala:566)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 5 in stage 17.0 failed 1 times, most recent failure: Lost task 5.0 in stage 17.0 (TID 7395, localhost, executor driver): org.apache.spark.SparkException: Failed to execute user defined function(anonfun$5: (struct<xmin:double,ymin:double,xmax:double,ymax:double>, struct<crsProj4:string>, array<struct<cell_context:struct<cellType:struct<cellTypeName:string>,dimensions:struct<cols:int,rows:int>>,cell_data:struct<cells:binary,ref:struct<source:struct<raster_source_kryo:binary>,bandIndex:int,subextent:struct<xmin:double,ymin:double,xmax:double,ymax:double>,subgrid:struct<colMin:int,rowMin:int,colMax:int,rowMax:int>>>>>, array<struct<xmin:double,ymin:double,xmax:double,ymax:double>>, array<struct<crsProj4:string>>, struct<cols:int,rows:int>) => struct<cell_context:struct<cellType:struct<cellTypeName:string>,dimensions:struct<cols:int,rows:int>>,cell_data:struct<cells:binary,ref:struct<source:struct<raster_source_kryo:binary>,bandIndex:int,subextent:struct<xmin:double,ymin:double,xmax:double,ymax:double>,subgrid:struct<colMin:int,rowMin:int,colMax:int,rowMax:int>>>>)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.ScalaUDF_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection$NestedClass_2.writeFields_930_221$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection$NestedClass_2.writeFields_930$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:435)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:441)
	at scala.collection.Iterator$JoinIterator.hasNext(Iterator.scala:212)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.apache.spark.sql.execution.UnsafeExternalRowSorter.sort(UnsafeExternalRowSorter.java:216)
	at org.apache.spark.sql.execution.SortExec$$anonfun$1.apply(SortExec.scala:108)
	at org.apache.spark.sql.execution.SortExec$$anonfun$1.apply(SortExec.scala:101)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.locationtech.rasterframes.functions.package$$anonfun$5.apply(package.scala:115)
	at org.locationtech.rasterframes.functions.package$$anonfun$5.apply(package.scala:100)
	... 27 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:167)
	... 33 more
Caused by: org.apache.spark.SparkException: Failed to execute user defined function(anonfun$5: (struct<xmin:double,ymin:double,xmax:double,ymax:double>, struct<crsProj4:string>, array<struct<cell_context:struct<cellType:struct<cellTypeName:string>,dimensions:struct<cols:int,rows:int>>,cell_data:struct<cells:binary,ref:struct<source:struct<raster_source_kryo:binary>,bandIndex:int,subextent:struct<xmin:double,ymin:double,xmax:double,ymax:double>,subgrid:struct<colMin:int,rowMin:int,colMax:int,rowMax:int>>>>>, array<struct<xmin:double,ymin:double,xmax:double,ymax:double>>, array<struct<crsProj4:string>>, struct<cols:int,rows:int>) => struct<cell_context:struct<cellType:struct<cellTypeName:string>,dimensions:struct<cols:int,rows:int>>,cell_data:struct<cells:binary,ref:struct<source:struct<raster_source_kryo:binary>,bandIndex:int,subextent:struct<xmin:double,ymin:double,xmax:double,ymax:double>,subgrid:struct<colMin:int,rowMin:int,colMax:int,rowMax:int>>>>)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.ScalaUDF_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection$NestedClass_2.writeFields_930_221$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection$NestedClass_2.writeFields_930$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:435)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:441)
	at scala.collection.Iterator$JoinIterator.hasNext(Iterator.scala:212)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.apache.spark.sql.execution.UnsafeExternalRowSorter.sort(UnsafeExternalRowSorter.java:216)
	at org.apache.spark.sql.execution.SortExec$$anonfun$1.apply(SortExec.scala:108)
	at org.apache.spark.sql.execution.SortExec$$anonfun$1.apply(SortExec.scala:101)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: java.lang.NullPointerException
	at org.locationtech.rasterframes.functions.package$$anonfun$5.apply(package.scala:115)
	at org.locationtech.rasterframes.functions.package$$anonfun$5.apply(package.scala:100)
	... 27 more

```

</details>

--- a/core/src/main/scala/org/locationtech/rasterframes/encoders/StandardSerializers.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/encoders/StandardSerializers.scala
@@ -49,7 +49,7 @@ trait StandardSerializers {
     ))
 
     override protected def to[R](t: Envelope, io: CatalystIO[R]): R = io.create(
-      t.getMinX, t.getMaxX, t.getMinY, t.getMaxX
+      t.getMinX, t.getMaxX, t.getMinY, t.getMaxY
     )
 
     override protected def from[R](t: R, io: CatalystIO[R]): Envelope = new Envelope(

--- a/core/src/main/scala/org/locationtech/rasterframes/expressions/DynamicExtractors.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/expressions/DynamicExtractors.scala
@@ -102,6 +102,41 @@ object DynamicExtractors {
       (v: Any) => v.asInstanceOf[InternalRow].to[CRS]
   }
 
+  /** This is necessary because extents created from Python Rows will reorder field names. */
+  object ExtentLike {
+
+    def rightShape(struct: StructType) =
+      struct.size == 4 && {
+        val n = struct.fieldNames.map(_.toLowerCase).toSet
+        n == Set("xmin", "ymin", "xmax", "ymax")|| n == Set("minx", "miny", "maxx", "maxy")
+      } && struct.fields.map(_.dataType).toSet == Set(DoubleType)
+
+
+    def unapply(dt: DataType): Option[Any => Extent] = dt match {
+      case dt: StructType if rightShape(dt) =>
+        Some((input: Any) => {
+          val row = input.asInstanceOf[InternalRow]
+
+          def maybeValue(name: String): Option[Double] = {
+            dt.indexWhere(_.name.toLowerCase == name) match {
+              case idx if idx >= 0 => Some(row.getDouble(idx))
+              case _ => None
+            }
+          }
+
+          def value(n1: String, n2: String): Double =
+            maybeValue(n1).orElse(maybeValue(n2)).getOrElse(throw new IllegalArgumentException(s"Missing field $n1 or $n2"))
+
+          val xmin = value("xmin", "minx")
+          val ymin = value("ymin", "miny")
+          val xmax = value("xmax", "maxx")
+          val ymax = value("ymax", "maxy")
+          Extent(xmin, ymin, xmax, ymax)
+        })
+      case _ => None
+    }
+  }
+
   lazy val extentExtractor: PartialFunction[DataType, Any ⇒ Extent] = {
     val base: PartialFunction[DataType, Any ⇒ Extent]= {
       case t if org.apache.spark.sql.rf.WithTypeConformity(t).conformsTo(JTSTypes.GeometryTypeInstance) =>
@@ -110,6 +145,7 @@ object DynamicExtractors {
         (input: Any) => input.asInstanceOf[InternalRow].to[Extent]
       case t if t.conformsTo[Envelope] =>
         (input: Any) => Extent(input.asInstanceOf[InternalRow].to[Envelope])
+      case ExtentLike(e) => e
     }
 
     val fromPRL = projectedRasterLikeExtractor.andThen(_.andThen(_.extent))

--- a/core/src/main/scala/org/locationtech/rasterframes/expressions/DynamicExtractors.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/expressions/DynamicExtractors.scala
@@ -96,10 +96,15 @@ object DynamicExtractors {
   }
 
   lazy val crsExtractor: PartialFunction[DataType, Any => CRS] = {
-    case _: StringType =>
-      (v: Any) => LazyCRS(v.asInstanceOf[UTF8String].toString)
-    case t if t.conformsTo[CRS] =>
-      (v: Any) => v.asInstanceOf[InternalRow].to[CRS]
+    val base: PartialFunction[DataType, Any ⇒ CRS] = {
+      case _: StringType =>
+        (v: Any) => LazyCRS(v.asInstanceOf[UTF8String].toString)
+      case t if t.conformsTo[CRS] =>
+        (v: Any) => v.asInstanceOf[InternalRow].to[CRS]
+    }
+
+    val fromPRL = projectedRasterLikeExtractor.andThen(_.andThen(_.crs))
+    fromPRL orElse base
   }
 
   /** This is necessary because extents created from Python Rows will reorder field names. */
@@ -138,7 +143,7 @@ object DynamicExtractors {
   }
 
   lazy val extentExtractor: PartialFunction[DataType, Any ⇒ Extent] = {
-    val base: PartialFunction[DataType, Any ⇒ Extent]= {
+    val base: PartialFunction[DataType, Any ⇒ Extent] = {
       case t if org.apache.spark.sql.rf.WithTypeConformity(t).conformsTo(JTSTypes.GeometryTypeInstance) =>
         (input: Any) => Extent(JTSTypes.GeometryTypeInstance.deserialize(input).getEnvelopeInternal)
       case t if t.conformsTo[Extent] =>

--- a/core/src/main/scala/org/locationtech/rasterframes/expressions/accessors/GetCRS.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/expressions/accessors/GetCRS.scala
@@ -27,12 +27,12 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
 import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
-import org.apache.spark.sql.types.{DataType, StringType}
+import org.apache.spark.sql.types.DataType
 import org.apache.spark.sql.{Column, TypedColumn}
 import org.apache.spark.unsafe.types.UTF8String
 import org.locationtech.rasterframes.encoders.CatalystSerializer._
 import org.locationtech.rasterframes.encoders.StandardEncoders.crsEncoder
-import org.locationtech.rasterframes.expressions.DynamicExtractors.projectedRasterLikeExtractor
+import org.locationtech.rasterframes.expressions.DynamicExtractors._
 import org.locationtech.rasterframes.model.LazyCRS
 
 /**
@@ -52,9 +52,8 @@ case class GetCRS(child: Expression) extends UnaryExpression with CodegenFallbac
   override def nodeName: String = "rf_crs"
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    if (child.dataType != StringType && !projectedRasterLikeExtractor.isDefinedAt(child.dataType)) {
-      TypeCheckFailure(s"Input type '${child.dataType}' does not conform to `String` or `ProjectedRasterLike`.")
-    }
+    if (!crsExtractor.isDefinedAt(child.dataType) )
+      TypeCheckFailure(s"Input type '${child.dataType}' does not conform to a CRS or something with one.")
     else TypeCheckSuccess
   }
 
@@ -62,8 +61,8 @@ case class GetCRS(child: Expression) extends UnaryExpression with CodegenFallbac
     input match {
       case s: UTF8String => LazyCRS(s.toString).toInternalRow
       case row: InternalRow ⇒
-        val prl = projectedRasterLikeExtractor(child.dataType)(row)
-        prl.crs.toInternalRow
+        val crs = crsExtractor(child.dataType)(row)
+        crs.toInternalRow
       case o ⇒ throw new IllegalArgumentException(s"Unsupported input type: $o")
     }
   }

--- a/core/src/main/scala/org/locationtech/rasterframes/extensions/RasterJoin.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/extensions/RasterJoin.scala
@@ -112,7 +112,7 @@ object RasterJoin {
 
     // Here's the meat:
     left
-      // 1. Add a unique ID to each LHS row for subequent grouping.
+      // 1. Add a unique ID to each LHS row for subsequent grouping.
       .withColumn(id, monotonically_increasing_id())
       // 2. Perform the left-outer join
       .join(right, joinExprs, joinType = "left")

--- a/core/src/main/scala/org/locationtech/rasterframes/extensions/RasterJoin.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/extensions/RasterJoin.scala
@@ -24,9 +24,8 @@ import org.apache.spark.sql._
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.types.DataType
 import org.locationtech.rasterframes._
-import org.locationtech.rasterframes.encoders.serialized_literal
-import org.locationtech.rasterframes.expressions.{DynamicExtractors, SpatialRelation}
 import org.locationtech.rasterframes.expressions.accessors.ExtractTile
+import org.locationtech.rasterframes.expressions.{DynamicExtractors, SpatialRelation}
 import org.locationtech.rasterframes.functions.reproject_and_merge
 import org.locationtech.rasterframes.util._
 
@@ -99,10 +98,9 @@ object RasterJoin {
 
     // After the aggregation we take all the tiles we've collected and resample + merge
     // into LHS extent/CRS.
-    // Use a representative tile from the left for the tile dimensions
-    val destDims = left.tileColumns.headOption
-      .map(t => rf_dimensions(unresolved(t)))
-      .getOrElse(serialized_literal(NOMINAL_TILE_DIMS))
+    // Use a representative tile from the left for the tile dimensions.
+    // Assumes all LHS tiles in a row are of the same size.
+    val destDims = rf_dimensions(coalesce(left.tileColumns.map(unresolved): _*))
 
     val reprojCols = rightAggTiles.map(t => {
       reproject_and_merge(

--- a/core/src/main/scala/org/locationtech/rasterframes/extensions/RasterJoin.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/extensions/RasterJoin.scala
@@ -22,9 +22,10 @@
 package org.locationtech.rasterframes.extensions
 import org.apache.spark.sql._
 import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.DataType
 import org.locationtech.rasterframes._
 import org.locationtech.rasterframes.encoders.serialized_literal
-import org.locationtech.rasterframes.expressions.SpatialRelation
+import org.locationtech.rasterframes.expressions.{DynamicExtractors, SpatialRelation}
 import org.locationtech.rasterframes.expressions.accessors.ExtractTile
 import org.locationtech.rasterframes.functions.reproject_and_merge
 import org.locationtech.rasterframes.util._
@@ -58,9 +59,18 @@ object RasterJoin {
     apply(left, right, joinExpr, leftExtent, leftCRS, rightExtent, rightCRS)
   }
 
+  private def checkType[T](col: Column, description: String, extractor: PartialFunction[DataType, Any => T]): Unit = {
+    require(extractor.isDefinedAt(col.expr.dataType), s"Expected column ${col} to be of type $description, but was ${col.expr.dataType}.")
+  }
+
   def apply(left: DataFrame, right: DataFrame, joinExprs: Column, leftExtent: Column, leftCRS: Column, rightExtent: Column, rightCRS: Column): DataFrame = {
     // Convert resolved column into a symbolic one.
     def unresolved(c: Column): Column = col(c.columnName)
+
+    checkType(leftExtent, "Extent", DynamicExtractors.extentExtractor)
+    checkType(leftCRS, "CRS", DynamicExtractors.crsExtractor)
+    checkType(rightExtent, "Extent", DynamicExtractors.extentExtractor)
+    checkType(rightCRS, "CRS", DynamicExtractors.crsExtractor)
 
     // Unique id for temporary columns
     val id = Random.alphanumeric.take(5).mkString("_", "", "_")
@@ -80,7 +90,7 @@ object RasterJoin {
     // On the LHS we just want the first thing (subsequent ones should be identical.
     val leftAggCols = left.columns.map(s => first(left(s), true) as s)
     // On the RHS we collect result as a list.
-    val rightAggCtx = Seq(collect_list(rightExtent) as rightExtent2, collect_list(rightCRS) as rightCRS2)
+    val rightAggCtx = Seq(collect_list(rightExtent) as rightExtent2, collect_list(rf_crs(rightCRS)) as rightCRS2)
     val rightAggTiles = right.tileColumns.map(c => collect_list(ExtractTile(c)) as c.columnName)
     val rightAggOther = right.notTileColumns
       .filter(n => n.columnName != rightExtent.columnName && n.columnName != rightCRS.columnName)
@@ -94,9 +104,11 @@ object RasterJoin {
       .map(t => rf_dimensions(unresolved(t)))
       .getOrElse(serialized_literal(NOMINAL_TILE_DIMS))
 
-    val reprojCols = rightAggTiles.map(t => reproject_and_merge(
-      col(leftExtent2), col(leftCRS2), col(t.columnName), col(rightExtent2), col(rightCRS2), destDims
-    ) as t.columnName)
+    val reprojCols = rightAggTiles.map(t => {
+      reproject_and_merge(
+        col(leftExtent2), col(leftCRS2), col(t.columnName), col(rightExtent2), col(rightCRS2), destDims
+      ) as t.columnName
+    })
 
     val finalCols = leftAggCols.map(unresolved) ++ reprojCols ++ rightAggOther.map(unresolved)
 

--- a/core/src/main/scala/org/locationtech/rasterframes/functions/package.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/functions/package.scala
@@ -109,18 +109,22 @@ package object functions {
       val rightExtents = rightExtentEnc.map(_.to[Extent])
       val rightCRSs = rightCRSEnc.map(_.to[CRS])
 
-      val cellType = tiles.map(_.cellType).reduceOption(_ union _).getOrElse(tiles.head.cellType)
+      if (leftExtent == null || leftDims == null || leftCRS == null) null
+      else {
 
-      // TODO: how to allow control over... expression?
-      val projOpts = Reproject.Options.DEFAULT
-      val dest: Tile = ArrayTile.empty(cellType, leftDims.cols, leftDims.rows)
-      //is there a GT function to do all this?
-      tiles.zip(rightExtents).zip(rightCRSs).map {
-        case ((tile, extent), crs) =>
-          tile.reproject(extent, crs, leftCRS, projOpts)
-      }.foldLeft(dest)((d, t) =>
-        d.merge(leftExtent, t.extent, t.tile, projOpts.method)
-      )
+        val cellType = tiles.map(_.cellType).reduceOption(_ union _).getOrElse(tiles.head.cellType)
+
+        // TODO: how to allow control over... expression?
+        val projOpts = Reproject.Options.DEFAULT
+        val dest: Tile = ArrayTile.empty(cellType, leftDims.cols, leftDims.rows)
+        //is there a GT function to do all this?
+        tiles.zip(rightExtents).zip(rightCRSs).map {
+          case ((tile, extent), crs) =>
+            tile.reproject(extent, crs, leftCRS, projOpts)
+        }.foldLeft(dest)((d, t) =>
+          d.merge(leftExtent, t.extent, t.tile, projOpts.method)
+        )
+      }
     }
   }
 

--- a/core/src/test/scala/org/locationtech/rasterframes/expressions/DynamicExtractorsSpec.scala
+++ b/core/src/test/scala/org/locationtech/rasterframes/expressions/DynamicExtractorsSpec.scala
@@ -1,0 +1,92 @@
+/*
+ * This software is licensed under the Apache 2 license, quoted below.
+ *
+ * Copyright 2020 Astraea, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *     [http://www.apache.org/licenses/LICENSE-2.0]
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.locationtech.rasterframes.expressions
+
+import geotrellis.vector.Extent
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.{Encoder, Encoders}
+import org.locationtech.jts.geom.Envelope
+import org.locationtech.rasterframes.TestEnvironment
+import org.locationtech.rasterframes.encoders.CatalystSerializer._
+import org.locationtech.rasterframes.expressions.DynamicExtractors._
+import org.locationtech.rasterframes.expressions.DynamicExtractorsSpec.{SnowflakeExtent1, SnowflakeExtent2}
+import org.scalatest.Inspectors
+
+class DynamicExtractorsSpec  extends TestEnvironment with Inspectors {
+  describe("Extent extraction") {
+    val expected = Extent(1, 2, 3, 4)
+    it("should handle normal Extent") {
+      extentExtractor.isDefinedAt(schemaOf[Extent]) should be(true)
+
+      val row = expected.toInternalRow
+      extentExtractor(schemaOf[Extent])(row) should be (expected)
+    }
+    it("should handle Envelope") {
+      extentExtractor.isDefinedAt(schemaOf[Envelope]) should be(true)
+
+      val e = expected.jtsEnvelope
+
+      val row = e.toInternalRow
+      extentExtractor(schemaOf[Envelope])(row) should be (expected)
+    }
+
+    it("should handle artisanally constructed Extents") {
+      // Tests the case where PySpark will reorder manually constructed fields.
+      // See https://stackoverflow.com/questions/35343525/how-do-i-order-fields-of-my-row-objects-in-spark-python/35343885#35343885
+
+      import spark.implicits._
+      withClue("case 1"){
+        val special = SnowflakeExtent1(expected.xmax, expected.ymin, expected.xmin, expected.ymax)
+        val df = Seq(Tuple1(special)).toDF("extent")
+        val encodedType = df.schema.fields(0).dataType
+        val encodedRow = SnowflakeExtent1.enc.toRow(special)
+        extentExtractor.isDefinedAt(encodedType) should be(true)
+        extentExtractor(encodedType)(encodedRow) should be(expected)
+      }
+
+      withClue("case 2") {
+        val special = SnowflakeExtent2(expected.xmax, expected.ymin, expected.xmin, expected.ymax)
+        val df = Seq(Tuple1(special)).toDF("extent")
+        val encodedType = df.schema.fields(0).dataType
+        val encodedRow = SnowflakeExtent2.enc.toRow(special)
+        extentExtractor.isDefinedAt(encodedType) should be(true)
+        extentExtractor(encodedType)(encodedRow) should be(expected)
+      }
+    }
+  }
+
+}
+
+object DynamicExtractorsSpec {
+  case class SnowflakeExtent1(xmax: Double, ymin: Double, xmin: Double, ymax: Double)
+
+  object SnowflakeExtent1 {
+    implicit val enc: ExpressionEncoder[SnowflakeExtent1] = Encoders.product[SnowflakeExtent1].asInstanceOf[ExpressionEncoder[SnowflakeExtent1]]
+  }
+
+  case class SnowflakeExtent2(xmax: Double, ymin: Double, xmin: Double, ymax: Double)
+
+  object SnowflakeExtent2 {
+    implicit val enc: ExpressionEncoder[SnowflakeExtent2] = Encoders.product[SnowflakeExtent2].asInstanceOf[ExpressionEncoder[SnowflakeExtent2]]
+  }
+
+}

--- a/core/src/test/scala/org/locationtech/rasterframes/expressions/DynamicExtractorsSpec.scala
+++ b/core/src/test/scala/org/locationtech/rasterframes/expressions/DynamicExtractorsSpec.scala
@@ -22,8 +22,8 @@
 package org.locationtech.rasterframes.expressions
 
 import geotrellis.vector.Extent
+import org.apache.spark.sql.Encoders
 import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
-import org.apache.spark.sql.{Encoder, Encoders}
 import org.locationtech.jts.geom.Envelope
 import org.locationtech.rasterframes.TestEnvironment
 import org.locationtech.rasterframes.encoders.CatalystSerializer._

--- a/pyrasterframes/src/main/python/pyrasterframes/__init__.py
+++ b/pyrasterframes/src/main/python/pyrasterframes/__init__.py
@@ -83,7 +83,7 @@ def _raster_join(df, other, left_extent=None, left_crs=None, right_extent=None, 
     else:
         jdf = ctx._jrfctx.rasterJoin(df._jdf, other._jdf)
 
-    return RasterFrameLayer(jdf, ctx._spark_session)
+    return DataFrame(jdf, ctx._spark_session)
 
 
 def _layer_reader(df_reader, format_key, path, **options):

--- a/pyrasterframes/src/main/python/pyrasterframes/rf_types.py
+++ b/pyrasterframes/src/main/python/pyrasterframes/rf_types.py
@@ -38,7 +38,7 @@ from pyrasterframes.rf_context import RFContext
 
 import numpy as np
 
-__all__ = ['RasterFrameLayer', 'Tile', 'TileUDT', 'CellType', 'Extent', 'RasterSourceUDT', 'TileExploder', 'NoDataFilter']
+__all__ = ['RasterFrameLayer', 'Tile', 'TileUDT', 'CellType', 'Extent', 'CRS', 'RasterSourceUDT', 'TileExploder', 'NoDataFilter']
 
 
 class cached_property(object):
@@ -221,16 +221,22 @@ class Extent(object):
         return self.__jvm__.toString()
 
 class CRS(object):
-    def __init__(self, proj4_str):
-        self.proj4_str = proj4_str
+    # NB: The name `crsProj4` has to match what's used in StandardSerializers.crsSerializers
+    def __init__(self, crsProj4):
+        self.crsProj4 = crsProj4
 
     @cached_property
     def __jvm__(self):
         comp = RFContext.active().companion_of("org.locationtech.rasterframes.model.LazyCRS")
-        return comp.apply(self.proj4_str)
+        return comp.apply(self.crsProj4)
 
     def __str__(self):
-        return self.proj4_str
+        return self.crsProj4
+
+    @property
+    def proj4_str(self):
+        """Alias for `crsProj4`"""
+        return self.crsProj4
 
 
 class CellType(object):

--- a/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
+++ b/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
@@ -434,6 +434,45 @@ class RasterJoin(TestEnvironment):
         with self.assertRaises(AssertionError):
             self.rf.raster_join(rf_prime, join_exprs=self.rf.extent)
 
+    def test_raster_join_with_null_left_head(self):
+        # https://github.com/locationtech/rasterframes/issues/462
+
+        from py4j.protocol import Py4JJavaError
+
+        ones = np.ones((10, 10), dtype='uint8')
+        e = Extent(0.0, 0.0, 40.0, 40.0)
+        c = 'EPSG:32611'
+
+        left = self.spark.createDataFrame(
+            [
+                Row(i=1, t=Tile(ones, CellType.uint8()), e=e, c=c),
+                Row(i=1, t=None, e=e, c=c)
+            ]
+        )
+
+        right = self.spark.createDataFrame(
+            [
+                Row(i=1, r=Tile(ones, CellType.uint8()), e=e, c=c),
+            ])
+
+        try:
+            joined = left.raster_join(right,
+                                      join_exprs=left.i == right.i,
+                                      left_extent=left.e, right_extent=right.e,
+                                      left_crs=left.c, right_crs=right.c)
+
+            self.assertEqual(joined.count(), 2)
+
+            collected = joined.select(rf_dimensions('r').cols.alias('cols'),
+                                      rf_dimensions('r').rows.alias('rows')) \
+                .collect()
+            for r in collected:
+                self.assertEqual(r.rows, 10)
+                self.assertEqual(r.cols, 10)
+
+        except Py4JJavaError as e:
+            self.fail('test_raster_join_with_null_left_head failed with Py4JJavaError:' + e)
+
 
 def suite():
     function_tests = unittest.TestSuite()

--- a/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
+++ b/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
@@ -440,13 +440,14 @@ class RasterJoin(TestEnvironment):
         from py4j.protocol import Py4JJavaError
 
         ones = np.ones((10, 10), dtype='uint8')
+        t = Tile(ones, CellType.uint8())
         e = Extent(0.0, 0.0, 40.0, 40.0)
         c = CRS('EPSG:32611')
 
         left = self.spark.createDataFrame(
             [
-                Row(i=1, t=Tile(ones, CellType.uint8()), e=e, c=c),
-                Row(i=1, t=None, e=e, c=c)
+                Row(i=1, j='a', t=t,    u=t, e=e, c=c),
+                Row(i=1, j='b', t=None, u=t, e=e, c=c)
             ]
         )
 
@@ -462,14 +463,38 @@ class RasterJoin(TestEnvironment):
                                       left_crs=left.c, right_crs=right.c)
 
             self.assertEqual(joined.count(), 2)
+            # In the case where the head column is null it will be passed thru
+            self.assertTrue(joined.select(isnull('t')).filter(col('j') == 'b').first()[0])
 
+            # The right hand side tile should get dimensions from col `u` however
             collected = joined.select(rf_dimensions('r').cols.alias('cols'),
                                       rf_dimensions('r').rows.alias('rows')) \
-                .dropna() \
                 .collect()
+
             for r in collected:
                 self.assertEqual(10, r.rows)
                 self.assertEqual(10, r.cols)
+
+            # If there is no non-null tile on the LHS then the RHS is ill defined
+            joined_no_left_tile = left.drop('u') \
+                .raster_join(right,
+                             join_exprs=left.i == right.i,
+                             left_extent=left.e, right_extent=right.e,
+                             left_crs=left.c, right_crs=right.c)
+            self.assertEqual(joined_no_left_tile.count(), 2)
+
+            # Tile col from Left side passed thru as null
+            self.assertTrue(
+                joined_no_left_tile.select(isnull('t')) \
+                    .filter(col('j') == 'b') \
+                    .first()[0]
+            )
+            # Because no non-null tile col on Left side, the right side is null too
+            self.assertTrue(
+                joined_no_left_tile.select(isnull('r')) \
+                    .filter(col('j') == 'b') \
+                    .first()[0]
+            )
 
         except Py4JJavaError as e:
             self.fail('test_raster_join_with_null_left_head failed with Py4JJavaError:' + e)

--- a/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
+++ b/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
@@ -441,7 +441,7 @@ class RasterJoin(TestEnvironment):
 
         ones = np.ones((10, 10), dtype='uint8')
         e = Extent(0.0, 0.0, 40.0, 40.0)
-        c = 'EPSG:32611'
+        c = CRS('EPSG:32611')
 
         left = self.spark.createDataFrame(
             [

--- a/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
+++ b/pyrasterframes/src/main/python/tests/PyRasterFramesTests.py
@@ -465,10 +465,11 @@ class RasterJoin(TestEnvironment):
 
             collected = joined.select(rf_dimensions('r').cols.alias('cols'),
                                       rf_dimensions('r').rows.alias('rows')) \
+                .dropna() \
                 .collect()
             for r in collected:
-                self.assertEqual(r.rows, 10)
-                self.assertEqual(r.cols, 10)
+                self.assertEqual(10, r.rows)
+                self.assertEqual(10, r.cols)
 
         except Py4JJavaError as e:
             self.fail('test_raster_join_with_null_left_head failed with Py4JJavaError:' + e)

